File: python/darkfactory/utils/system.py
"""System-level utility checks."""

from __future__ import annotations

import shutil
from pathlib import Path


def check_prerequisites(cwd: Path, *, require_claude: bool = True) -> None:
    """Fail fast if required tools are missing or cwd is not a git repo."""
    if require_claude and shutil.which("claude") is None:
        raise SystemExit(
            "error: 'claude' is not on PATH. Install Claude Code to use 'prd discuss'."
        )
    if shutil.which("git") is None:
        raise SystemExit("error: 'git' is not on PATH.")
    if not (cwd / ".git").exists():
        # Walk up to check if we're inside a git worktree (avoids importing
        # from cli, which would be an inverted dependency).
        current = cwd.resolve()
        while current != current.parent:
            if (current / ".git").exists():
                break
            current = current.parent
        else:
            raise SystemExit(
                "error: current directory is not inside a git working tree."
            )

File: ts/src/utils/exec/system.ts
/**
 * system.ts — prerequisite checks.
 *
 * Returns Result instead of throwing SystemExit like the Python version.
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { type Result, err, ok } from "../result.js";
import { which } from "./subprocess.js";

export interface PrerequisiteErr {
  readonly kind: "prerequisite-err";
  readonly missing: readonly string[];
  readonly message: string;
}

/**
 * Check that git, gh, and optionally claude are on PATH, and that
 * cwd is inside a git working tree.
 *
 * Returns Result<null, PrerequisiteErr> — no exceptions.
 */
export function checkPrerequisites(
  cwd: string,
  options?: { requireClaude?: boolean }
): Result<null, PrerequisiteErr> {
  const missing: string[] = [];

  if (which("git") === null) {
    missing.push("git");
  }

  if (which("gh") === null) {
    missing.push("gh");
  }

  if ((options?.requireClaude ?? true) && which("claude") === null) {
    missing.push("claude");
  }

  if (missing.length > 0) {
    const list = missing.map((m) => `'${m}'`).join(", ");
    return err({
      kind: "prerequisite-err",
      missing,
      message: `error: required tools not on PATH: ${list}`,
    });
  }

  // Check that cwd is inside a git working tree
  if (!isInsideGitWorkTree(cwd)) {
    return err({
      kind: "prerequisite-err",
      missing: [],
      message: "error: current directory is not inside a git working tree.",
    });
  }

  return ok(null);
}

/**
 * Walk up directory tree to find a .git file or directory.
 */
function isInsideGitWorkTree(startDir: string): boolean {
  let current = path.resolve(startDir);
  const root = path.parse(current).root;

  while (current !== root) {
    const gitPath = path.join(current, ".git");
    try {
      fs.accessSync(gitPath);
      return true;
    } catch {
      // Not found here — walk up
    }
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
  }

  return false;
}
